### PR TITLE
azure-pipelines: more complete port of travis conf

### PR DESCRIPTION
**Pull request by ethomson**
## Overview

A more complete port of the Travis configuration.  I talked with @brunoborges about porting more of the CI invocation over to Azure Pipelines; this picks up from his #1796 with a few differences:

1. For the Java 11 (Linux), macOS and Windows-based builds, this runs directly on the Azure Pipelines build agents, since they have Java 11 installed.
2. For the Java 12 and Java 13 based builds, this runs within the OpenJDK docker images.  The Azure Pipelines build agents do not have a Java 12 or 13 runtime installed, so the preferred mechanism is to [run within a docker image](https://devblogs.microsoft.com/devops/cross-platform-container-builds-with-azure-pipelines/).
3. Just run gradle on the command line instead of trying to run the gradle task.  This allows us to use the same script configuration and template for both the container-based builds and the on-agent builds.
4. Only one of the builds (I chose the Java 12 Linux build) publishes test results, per @sormuras's request in #1796.

Outstanding Issues:

1. The publish steps do _not_ succeed.  (Thankfully.)  They require authentication which obviously I did not configure, as I do not have that information.  I'm happy to help configure secret variables so that we can move that forward.

Please let me know how else I can help!  😀 

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [ ] There are no TODOs left in the code
- [ ] Method [preconditions](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [ ] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [ ] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [ ] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [ ] Change is documented in the [User Guide](http://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](http://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [ ] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

**Comment by ethomson** [comment]
(Also, 👋.  I'm @ethomson from the Azure Pipelines team and a long-time junit user.  @brunoborges asked me to help out with some of the translation, and to go ahead and open this pull request since he's on holiday this week.)

**Comment by sormuras** [comment]
Hi @ethomson. Thanks for the PR, very much appreciated! Scanning it for details and already impressed, that the `Linux (OpenJDK 13)` finished after 3 minutes. :rocket:

File: azure-pipelines.yml [hunk 1]
```diff
@@ -1,23 +1,89 @@
...
+
+- job: linux_13
+  displayName: Linux (OpenJDK 13)
+  pool:
+    vmImage: 'ubuntu-16.04'
+  container: openjdk:13
+  steps:
+    - template: src/azure/azure-gradle-step.yml
+      parameters:
+        continueOnError: true
```
Comment (sormuras) [inline azure-pipelines.yml]: Ah, here it says: `allow failures`

**Comment by ethomson** [comment]
> OpenJDK 13 fails due to Groovy not being there, yet. 🙈

Yeah, that was a limitation that I elected not to address with this PR.  😉

File: src/azure/azure-gradle-step.yml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+        export JAVA_HOME="${JAVA_HOME_11_X64}"
+      fi
+      ./gradlew --version
+      ./gradlew --scan --warning-mode=all -Dplatform.tooling.support.tests.enabled=true build
+    displayName: Test
+    continueOnError: ${{ parameters.continueOnError }}
+  - task: PublishTestResults@2
+    displayName: Publish Test Results
+    inputs:
+      testResultsFiles: '**/build/test-results/test/TEST-*.xml'
```
Comment (sormuras) [inline src/azure/azure-gradle-step.yml]: Wondering why this also catches the "expected to fail" reports from the "documentation" module...

...either we need to exclude those, or delete the reports before collecting test results.

Comment (marcphilipp) [reply src/azure/azure-gradle-step.yml]: Actually `com.example.vintage.VintageTest` is from the platform-tooling-support-tests project.

@ethomson Can we exclude '**/build/test-workspace/**'?

Comment (sormuras) [reply src/azure/azure-gradle-step.yml]: Indeed. If excluding doesn't work out, I'll delete those reports after `platform-tooling-support-tests` are performed.

/fup2 #1810

**Comment by ethomson** [comment]
> One question: What does eq(variables['Build.Reason'], 'IndividualCI') do?

@marcphilipp It's a condition for a build that is queued because the target (master, maint, etc) branch was pushed into (or had a pull request merged).  As opposed to a pull request validation build or a build queued through the API.  (Why "individual" you might ask?  Azure Pipelines also has the notion of batching multiple changes for repositories with a high commit velocity or slow build times; if your build was a batch of commits, it wouldn't be an "individual" CI build.  This is rarely used.)

File: src/azure/azure-gradle-step.yml (hunk 2 above)
Comment (ethomson) [reply src/azure/azure-gradle-step.yml]: @marcphilipp @sormuras Unfortunately `testResultsFiles` only takes wildcards and there's no exclusion pattern.  If `rm` is okay then that's the easiest path forward, tbqh.

Comment (sormuras) [reply src/azure/azure-gradle-step.yml]: Already fixed via https://github.com/junit-team/junit5/commit/7f535408b182cb2ad2d5befe1a62c928439b987a

**Comment by marcphilipp** [comment]
Thanks for the explanation!

**Pull request merged**